src/excmds/open.ts

```diff
@@ -0,0 +1,6 @@
+import * as excmd_lib from "@src/lib/excmd_lib"
+import * as _excmds from "@src/excmds/_open"
+type ft = typeof _excmds
```

Comment:
Is there any way we can move this repeated code into a macro or function? Looks like boilerplate to me.

Reply:
If you're talking about consolidating them into a single file, I want to leave the excmds broken out; I've realized while working on this that Tridactyl has this huge blob of excmds that lets everything talk to every single excmd and it's nearly impossible to figure out what uses what. Breaking them out into modules means better encapsulation, which in turn means we're more able to refactor when the time comes.

If you're talking about just reducing the copy-paste... yeah, I'll look into that. I don't know what it'll look like, though; it's a good candidate for generated code, but I don't like having a bunch of generated code actually on disk. Maybe make every file with actual code in it use explicit exports so I can wrap them at the very end? Some way to tell webpack or the compiler to mangle the file as it's imported?

Reply:
I'm mostly talking about reducing the copy-paste.

Perhaps a compromise would be a python script to add a new excmd file, and all the associated gubbins (e.g. message types).

Another thing - does this PR break the help page into lots of little ones? I think that's one of the nicest parts of Tridactyl. I guess we might just have to write our own/fork TypeDoc quite heavily.

Reply:
I've been thinking about rewriting the help pages in React (or any other UI framework) for a few months now. We can use the information in .metadata.generated.ts to do that. I think it would have a lot of advantages:
- it would allow us to get rid of one compilation step
- it reduces the number of dependencies we have
- it gives us more freedom layout/presentation wise
- it lets us remove the awful hacks I wrote in `src/help.ts`
- it'd be trivial to replace the data generated by the custom typescript pass with the data we would get if we move to the `declareBackgroundFunction(name, docstring, args, function)` architecture I mentioned on IRC.
- It would provide a way for end-users to insert their own help items in the help pages, which would be a great step toward making a plugin API

Reply:
@bovine3dom This PR does _not_ split the help page up. The help page is constructed from the signature and docstring of the functions in `exmcds.ts`. The only functions that I've actually removed had `@hidden` in their docstrings and didn't show up in the help anyway. Every excmd is still present as an excmd and has just been refactored so it's a one-liner calling the library function with its functionality.

> I've been thinking about rewriting the help pages in React (or any other UI framework) for a few months now. We can use the information in .metadata.generated.ts to do that. I think it would have a lot of advantages:

Strongly agree that a help page using the generated metadata would be a major improvement.

I think that `declareBackgroundFunction` or some similar call would be a significant improvement over the current system of extracting type information by string-mangling the function signature. It's exactly the kind of streamlining I was talking about above when I said that it'd be really nice if excmds weren't being used as common library functions from typescript.

Reply:
...Huh. I have a new idea. Give me a bit, new PR will be going up that takes this in a slightly different direction.

Reply:
Try this: https://github.com/tridactyl/tridactyl/pull/1506

Reply:
Ah, I see. My initial reaction is that it isn't very ergonomic - I'd really like adding an ex-cmd to be as easy as possible. It seems like those "shims" could be generated.

Reply:
Ah. Well... As it stands, the bare minimum for implementing an excmd requires:
1. A declaration that an excmd exists with a particular name
2. A signature Documentation
4. Entry point to implementation

You need a name to make it part of the excmd language. You need a signature and documentation for the help page, and you need a signature so you can translate from excmd language to function-call. Finally, you need the entry-point to the implementation; essentially, this is the low-level side of a foreign function interface boundary between the excmd language and the typescript libraries that power Tridactyl.

Right now, we get these from:
1. The existence of a function in `excmds.ts` with `//#content` or `//#background` annotations (as opposed to `//#content_helper` or `//#background_helper`).
2. The signature of that function, subject to constraints.
3. The documentation comment for that function, as long as it doesn't have `/** @hidden */` in its doc-comment.
4. Function body, lol.

The thing is that none of these necessarily involve the actual implementation. Say that we move all of the actual _code_ into libs. `excmds.ts` will have nothing in it but function names, function signatures, and documentation comments - or, more concisely, nearly-minimal excmd declarations. With nothing left but those declarations, we can come up with some better format for declarations (goodbye `excmd_macros.py`) and change to use that. I explicitly _want_ to introduce a very small amount of "redundancy" here because that redundancy gives us breathing room with which to execute a significant refactor. Think activation energy - we're in a local minimum, we need to bump just a touch to get into a global minimum elsewhere.

If you're really worried about that layer of redundancy... how often do we really change the signatures or documentation of excmds? When we do do that, how often is it the case that the change is entirely contained in the body of the affected function in `excmds.ts`? Maybe I've just been setting my sights on bigger chunks of functionality, but it's already the case that we're having to keep documentation in sync with code changes elsewhere in the project. Which is entirely predictable.

As for adding an excmd being easy - I think that the complexity of declaring an excmd is going to be dwarfed by the complexity of implementing the exmcd for all but the simplest of implementations with the most hideously subtle documentable behaviors. Especially once we hollow out `excmts.ts` and accomplish a refactor. Right now with the wacky `//#content` and `//#background` and `//#content_helper` stuff and the string-mangling signature extraction. In fact, that signature extraction I think is _by far_ the worst part of implementing an excmd; I haven't found any documentation on what types are acceptable (e.g. "no types, you have to inline the string type literals") for arguments or how conversions from excmd string literals to those types. I'd really prefer to get rid of all that [magic](https://en.wikipedia.org/wiki/Magic_(programming)), and I think that the way to do that is to decouple excmd-implementation from the excmd-declaration so that we can safely change one without touching the other.

Reply:
I don't deny that the current method is slightly opaque - I'm just keen to keep the barrier to entry for new contributors as low as possible. The current process for adding an ex-cmd is:

- copy and paste one that looks similar to the one you want to make
- change the name, documentation, and the body to do what you want

I think I'd be happier with a method that needed multiple files to be edited if we had some tooling for adding skeletons for a new ex cmd. It could also be the case that we've matured enough that we don't need many more ex-cmds, so perhaps we need to find a new on-ramp for contributors.

Reply:
I do feel like we've explored a lot of the design space for simple excmds. Going down the issue tracker, I'm finding like two issues in the first 8 pages that'd want an excmd (#956, #1449), and none of those are anything _close_ to being good first issues.

Let me fiddle with some options for declaring excmds more easily.

Reply:
Also, I'd like to point out that this approach only requires two files to be modified. You declare your excmd in `exmcds.ts`, and you add the implementation in either `content_auto` or `background_auto`. If you want, I can add some stuff to the onboarding documentation about that?

Reply:
Ugh. Using typedoc to build `:help` means that we *have* to have a function for every excmd. That's... disappointing.

Reply:
I can't remember how we chose TypeDoc, but I strongly suspect our (my?) main criterion was maximal impact for minimum effort in the short term. I'm happy for it to be replaced - we could always leave config etc on TypeDoc while excmds got something new.

Reply:
Though... I don't know if there's anything that says that that function has to be the actual entry point, so we could still move to a more declarative `excmds.ts` by building a script that loads all of the declarations and generates a skeleton with empty functions with appropriate signatures for typedoc...

That's kind of disgusting. I should go look into documentation tools that are less tightly coupled to the source language. Or maybe I can find some hook into typedoc that will let me convince it that an excmd declaration is actually a function...

Reply:
Added some documentation. Better?

Reply:
Yes, documentation is good.

Some more bike-shedding though: I don't understand why the generated files are in version control, and I think they should be more obviously generated, e.g. /lib/generated/. I think background_auto is also a little misleading, perhaps background_src would be better?

Reply:
> I don't understand why the generated files are in version control

...Oops. Apparently I'm stupid.

> background_src

`background_lib`?

Reply:
Done.